### ISSUE TITLE
Expand Storage

### ISSUE THREAD

**Issue by aiesam** [reactive-tech/kubegres#34]
I just started using the operator and I think its got a really good potential.  How do you expand the storage class for new Pods in the same Statefulset without affecting existing pods?

**Comment by alex-arica** [comment]
Thank you for your message.

I suppose by saying expand the storage class, you mean increasing the size of PVC allocated to each Pod?

In the YAML you can modify the field 'spec.database.size' to a higher value. 

Kubegres will then check if the storage class has the field 'AllowVolumeExpansion' set to true. If that's the case, it will update the associated PVC and change their size.  It will start by the Replica instances and if it works it will propagate the changes to the Primary instance.

That will apply for all Pods. 

I hope that helps?

**Comment by aiesam** [comment]
Thanks Alex. That is exactly what I did, but get the error when trying to increase from 150Mi to 250Mi and also have ALLOWVOLUMEEXPANSION set to true. Not sure what I am missing.

 ```
 Normal  StatefulSetOperation          27s                   Kubegres-controller  Updated a StatefulSet with up-to-date Spec. 'StatefulSet name': my-kubegres-5, 'SpecName': StorageClassSize, 'New': 250Mi
  Normal  StatefulSetOperation          27s                   Kubegres-controller  Enforced the Spec of a StatefulSet. 'StatefulSet name': my-kubegres-5
  Normal  StatefulSetOperation          5s (x2 over 27s)      Kubegres-controller  The Spec is NOT up-to-date for a StatefulSet. 'StatefulSet name': my-kubegres-5, 'SpecName': StorageClassSize, 'Expected': 250Mi, 'Current': 150Mi

```

```
NAME                   PROVISIONER                                   RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
expanded-nfs-storage   k8s-sigs.io/nfs-subdir-external-provisioner   Delete          Immediate           true                   98m

```

**Comment by alex-arica** [comment]
Thank you for those details.

The messages from Kubegres do not seem to be showing an error. They are notifying that the StatefulSet is going to be updated. What do you get as messages after that?

Do you see any warning or error?

And what do you see for the PVC as events (with kubectl describe)? Was their size updated?

Thank you.

**Comment by aiesam** [comment]
No. The PVC was not updated. Itried again to change from 250Mi to 300Mi and The controller logs says

`2021-08-25T13:07:19.721Z        DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"Kubegres","namespace":"default","name":"my-kubegres","uid":"f8d9f6f1-7bf7-47fe-a49b-c8c66ba94029","apiVersion":"kubegres.reactive-tech.io/v1","resourceVersion":"2842820"}, "reason": "StatefulSetOperation", "message": "The Spec is NOT up-to-date for a StatefulSet. 'StatefulSet name': my-kubegres-5, 'SpecName': StorageClassSize, 'Expected': 300Mi, 'Current': 250Mi"}`

In addition, when that resizing operation fails, it block other operations such such as scaling the pod up or down.

**Comment by aiesam** [comment]
This exception is also from the controller log.

```
2021-08-25T13:10:13.756Z        INFO    controllers.Kubegres    The Spec is NOT up-to-date for a StatefulSet.   {"StatefulSet name": "my-kubegres-5", "SpecName": "StorageClassSize", "Expected": "300Mi", "Current": "250Mi"}
2021-08-25T13:10:13.756Z        ERROR   controllers.Kubegres    Last Spec enforcement attempt has timed-out for a StatefulSet. You must apply different spec changes to your Kubegres resource since the previous spec changes did not work. Until you apply it, most of the features of Kubegres are disabled for safety reason.     {"StatefulSet's name": "my-kubegres-5", "One or many of the following specs failed: ": "StorageClassSize: 300Mi", "error": "Spec enforcement timed-out"}
github.com/go-logr/zapr.(*zapLogger).Error
        /go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
reactive-tech.io/kubegres/controllers/ctx/log.(*LogWrapper).Error
        /workspace/controllers/ctx/log/LogWrapper.go:49
reactive-tech.io/kubegres/controllers/ctx/log.(*LogWrapper).ErrorEvent
        /workspace/controllers/ctx/log/LogWrapper.go:62
reactive-tech.io/kubegres/controllers/spec/enforcer/statefulset_spec.(*AllStatefulSetsSpecEnforcer).logSpecEnforcementTimedOut
        /workspace/controllers/spec/enforcer/statefulset_spec/AllStatefulSetsSpecEnforcer.go:169
reactive-tech.io/kubegres/controllers/spec/enforcer/statefulset_spec.(*AllStatefulSetsSpecEnforcer).EnforceSpec
        /workspace/controllers/spec/enforcer/statefulset_spec/AllStatefulSetsSpecEnforcer.go:109
reactive-tech.io/kubegres/controllers.(*KubegresReconciler).enforceAllStatefulSetsSpec
        /workspace/controllers/kubegres_controller.go:146
reactive-tech.io/kubegres/controllers.(*KubegresReconciler).enforceSpec
        /workspace/controllers/kubegres_controller.go:138
reactive-tech.io/kubegres/controllers.(*KubegresReconciler).Reconcile
        /workspace/controllers/kubegres_controller.go:100
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:263
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:235
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:198
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.UntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99
2021-08-25T13:10:13.756Z        DEBUG   controller-runtime.manager.events       Warning {"object": {"kind":"Kubegres","namespace":"default","name":"my-kubegres","uid":"f8d9f6f1-7bf7-47fe-a49b-c8c66ba94029","apiVersion":"kubegres.reactive-tech.io/v1","resourceVersion":"2842820"}, "reason": "StatefulSetSpecEnforcementTimedOutErr", "message": "Last Spec enforcement attempt has timed-out for a StatefulSet. You must apply different spec changes to your Kubegres resource since the previous spec changes did not work. Until you apply it, most of the features of Kubegres are disabled for safety reason.  'StatefulSet's name': my-kubegres-5, 'One or many of the following specs failed: ': StorageClassSize: 300Mi - Spec enforcement timed-out"}
2021-08-25T13:10:13.756Z        DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"Kubegres","namespace":"default","name":"my-kubegres","uid":"f8d9f6f1-7bf7-47fe-a49b-c8c66ba94029","apiVersion":"kubegres.reactive-tech.io/v1","resourceVersion":"2842820"}, "reason": "BlockingOperationTimedOut", "message": "Blocking-Operation timed-out. 'OperationId': Enforcing StatefulSet's Spec, 'StepId': StatefulSet's spec is updating"}
2021-08-25T13:10:13.756Z        DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"Kubegres","namespace":"default","name":"my-kubegres","uid":"f8d9f6f1-7bf7-47fe-a49b-c8c66ba94029","apiVersion":"kubegres.reactive-tech.io/v1","resourceVersion":"2842820"}, "reason": "StatefulSetOperation", "message": "The Spec is NOT up-to-date for a StatefulSet. 'StatefulSet name': my-kubegres-5, 'SpecName': StorageClassSize, 'Expected': 300Mi, 'Current': 250Mi"}

```

**Comment by alex-arica** [comment]
Thank you for those details. 

What seems to happen is the PVC is not responding to an update operation from Kubegres to set a new size. It could be an issue with the underlying storage class or something else which needs to be done for that specific storage clas

I need to use NFS and reproduce this issue locally first before being able to tell you the reason. 

Note that if you set again the old size in your YAML, it should work again.

**Comment by aiesam** [comment]
I believe this issue is due to the underlying nfs storage. I will close this case.

**Comment by alex-arica** [comment]
I think it could be an issue with the way Kubegres tries to expands the storage. 

I am going to work on this potential bug fix next.

**Comment by alex-arica** [comment]
I committed a potential bug fix in master.

To test it, please run the following command in your Kubernetes cluster:

```
kubectl apply -f https://raw.githubusercontent.com/reactive-tech/kubegres/main/kubegres.yaml
```

Could you please help me by testing it and let me know if you are happy with it?

Once you confirmed it is working for you, I will create a new tag 1.10

Please let me know and thanks in advance for your help!

**Comment by edwardzjl** [comment]
> I committed a potential bug fix in master.
> 
> To test it, please run the following command in your Kubernetes cluster:
> 
> ```
> kubectl apply -f https://raw.githubusercontent.com/reactive-tech/kubegres/main/kubegres.yaml
> ```
> 
> Could you please help me by testing it and let me know if you are happy with it?
> 
> Once you confirmed it is working for you, I will create a new tag 1.10
> 
> Please let me know and thanks in advance for your help!


I got another error message after applying the yaml:

```
2021-09-15T04:29:50.688Z        INFO    controllers.Kubegres    Updating Spec of a StatefulSet 
{"StatefulSet name": "postgres-3"}
2021-09-15T04:29:50.694Z        ERROR   controllers.Kubegres    Unable to enforce the Spec of a StatefulSet.   {"StatefulSet name": "postgres-3", "error": "StatefulSet.apps \"postgres-3\" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden"}
github.com/go-logr/zapr.(*zapLogger).Error
        /go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
reactive-tech.io/kubegres/controllers/ctx/log.(*LogWrapper).Error
        /workspace/controllers/ctx/log/LogWrapper.go:49
reactive-tech.io/kubegres/controllers/ctx/log.(*LogWrapper).ErrorEvent
        /workspace/controllers/ctx/log/LogWrapper.go:62
reactive-tech.io/kubegres/controllers/spec/enforcer/statefulset_spec.(*AllStatefulSetsSpecEnforcer).enforceSpec
        /workspace/controllers/spec/enforcer/statefulset_spec/AllStatefulSetsSpecEnforcer.go:189
reactive-tech.io/kubegres/controllers/spec/enforcer/statefulset_spec.(*AllStatefulSetsSpecEnforcer).EnforceSpec
        /workspace/controllers/spec/enforcer/statefulset_spec/AllStatefulSetsSpecEnforcer.go:119
reactive-tech.io/kubegres/controllers.(*KubegresReconciler).enforceAllStatefulSetsSpec
        /workspace/controllers/kubegres_controller.go:146
reactive-tech.io/kubegres/controllers.(*KubegresReconciler).enforceSpec
        /workspace/controllers/kubegres_controller.go:138
reactive-tech.io/kubegres/controllers.(*KubegresReconciler).Reconcile
        /workspace/controllers/kubegres_controller.go:100
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:263
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem       
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:235
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:198
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.UntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99
2021-09-15T04:29:50.694Z        INFO    controllers.Kubegres    Updating Kubegres' status:
 {"Field": "BlockingOperation", "New value": {"statefulSetOperation":{},"statefulSetSpecUpdateOperation":{}}}
2021-09-15T04:29:50.694Z        INFO    controllers.Kubegres    Updating Kubegres' status:
 {"Field": "PreviousBlockingOperation", "New value": {"operationId":"Enforcing StatefulSet's Spec","stepId":"StatefulSet's spec is updating","timeOutEpocInSeconds":1631680490,"statefulSetOperation":{"instanceIndex":3,"name":"postgres-3"},"statefulSetSpecUpdateOperation":{"specDifferences":"StorageClassSize: 10Gi"}}}
2021-09-15T04:29:50.694Z        DEBUG   controller-runtime.manager.events       Warning {"object": {"kind":"Kubegres","namespace":"kubegres-system","name":"postgres","uid":"d975e85d-736c-4d26-99b2-b163bbee1b62","apiVersion":"kubegres.reactive-tech.io/v1","resourceVersion":"57255268"}, "reason": "StatefulSetSpecEnforcementErr", "message": "Unable to enforce the Spec of a StatefulSet. 'StatefulSet name': postgres-3 - StatefulSet.apps \"postgres-3\" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden"}
```

**Comment by alex-arica** [comment]
@edwardzjl  Thank you for those details.

I am going to make a new change tomorrow and see if it resolves it. 

By the way, which local Kubernetes cluster are you using?

**Comment by edwardzjl** [comment]
@alex-arica I'm deploying kubegres on aliyun ecs with 2 master nodes and 5 worker nodes, with kubernetes version vary from 1.21.0 to 1.21.2
And the storage was managed by rook ceph.

**Comment by alex-arica** [comment]
@edwardzjl thank you for those details.

**Comment by alex-arica** [comment]
To test my changes locally, I am running Kubernetes with [Kind](https://kind.sigs.k8s.io/). The provisioner used by Kind ```rancher.io/local-path``` does not support volume expansion.

I did some research and I could not find any provisioners supporting volume expansion on the local hard disk.

I found [open-local](https://github.com/alibaba/open-local/blob/main/docs/user-guide.md) which supports volume expansion but I had to raise an issue https://github.com/alibaba/open-local/issues/47 as I am having some issues.

I have also created an issue https://github.com/kubernetes-sigs/kind/issues/2469 with Kind and asked to them whether they could recommend any provisioner. 

Meanwhile, if you have any suggestion, please let me know.

**Comment by alex-arica** [comment]
@edwardzjl , @aiesam Could you please do me a favour by installing Kubegres 1.10 and testing the increase of storage size?

I made a change in Kubegres 1.10 so that when the storage's size is changed, I force the restart of StatefulSets and Pods so that they pick-up the new storage size from their PVC. 

To install and test Kubegres 1.10, please run:
```
kubectl apply -f https://raw.githubusercontent.com/reactive-tech/kubegres/v1.10/kubegres.yaml
```

**Comment by edwardzjl** [comment]
I installed 1.10 (upgraded from 1.9 by kubectl apply) on my cluster

I then created a demo pg cluster with 200M as initial storage size and expand to 300M:

```yaml
apiVersion: kubegres.reactive-tech.io/v1
kind: Kubegres
metadata:
  name: demo-postgres
  namespace: demo-pg
spec:
   replicas: 3
   image: postgres:13.2
   database:
      size: 200Mi # to 300Mi
   env:
      - name: POSTGRES_PASSWORD
        valueFrom:
           secretKeyRef:
              name: postgres-secret
              key: superUserPassword
      - name: POSTGRES_REPLICATION_PASSWORD
        valueFrom:
           secretKeyRef:
              name: postgres-secret
              key: replicationUserPassword

```

After some waiting, only the 3rd pod got restarted:

![image](https://user-images.githubusercontent.com/7287580/134452212-664fc993-c2f9-4f0b-8881-156b0cb3fb3f.png)


I checked the pvc and pv, only the 3rd pod's volume got expanded:

![image](https://user-images.githubusercontent.com/7287580/134452318-87c9c283-7b8b-4fe7-bb05-571bc7b4b7d9.png)

and the controller log keeps looping:

```
2021-09-23T03:46:25.155Z        INFO    controllers.Kubegres    =======================================================
2021-09-23T03:46:25.155Z        INFO    controllers.Kubegres    =======================================================
2021-09-23T03:46:26.155Z        INFO    controllers.Kubegres    KUBEGRES        {"name": "demo-postgres", "Status": {"lastCreatedInstanceIndex":3,"blockingOperation":{"operationId":"Enforcing StatefulSet's Spec","stepId":"StatefulSet's spec is updating","timeOutEpocInSeconds":1632368833,"statefulSetOperation":{"instanceIndex":3,"name":"demo-postgres-3"},"statefulSetSpecUpdateOperation":{"specDifferences":"StorageClassSize: 300Mi"}},"previousBlockingOperation":{"operationId":"Replica DB count spec enforcement","stepId":"Replica DB is deploying","timeOutEpocInSeconds":1632368754,"statefulSetOperation":{"instanceIndex":3,"name":"demo-postgres-3"},"statefulSetSpecUpdateOperation":{}},"enforcedReplicas":3}}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    The Spec is NOT up-to-date for a StatefulSet.   {"StatefulSet name": "demo-postgres-3", "SpecName": "StorageClassSize", "Expected": "300Mi", "Current": "200Mi"}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    Active Blocking-Operation       {"OperationId": "Enforcing StatefulSet's Spec", "StepId": "StatefulSet's spec is updating", "HasTimedOut": false, "StatefulSetInstanceIndex": 3, "StatefulSetSpecDifferences": "StorageClassSize: 300Mi", "NbreSecondsLeftBeforeTimeOut": 47}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    Previous Blocking-Operation     {"OperationId": "Replica DB count spec enforcement", "StepId": "Replica DB is deploying", "HasTimedOut": false, "StatefulSetInstanceIndex": 3}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    Database StorageClass states.   {"IsDeployed": true, "name": "rook-ceph-block"}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    Base Config states      {"IsDeployed": true, "name": "base-kubegres-config"}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    All StatefulSets deployment states:     {"Spec expected to deploy": 3, "Nbre Deployed": 3}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    Primary states:         {"IsDeployed": true, "Name": "demo-postgres-1", "IsReady": true, "Pod Name": "demo-postgres-1-0", "Pod IsDeployed": true, "Pod IsReady": true, "Pod IsStuck": false}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    Replica states:         {"IsDeployed": true, "Name": "demo-postgres-2", "IsReady": true, "Pod Name": "demo-postgres-2-0", "Pod IsDeployed": true, "Pod IsReady": true, "Pod IsStuck": false}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    Replica states:         {"IsDeployed": true, "Name": "demo-postgres-3", "IsReady": true, "Pod Name": "demo-postgres-3-0", "Pod IsDeployed": true, "Pod IsReady": true, "Pod IsStuck": false}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    Primary Service states:         {"IsDeployed": true, "name": "demo-postgres"}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    Replica Service states:         {"IsDeployed": true, "name": "demo-postgres-replica"}
2021-09-23T03:46:26.156Z        INFO    controllers.Kubegres    BackUp states.  {"IsCronJobDeployed": false, "IsPvcDeployed": false, "CronJobLastScheduleTime": ""}
2021-09-23T03:46:26.156Z        DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"Kubegres","namespace":"demo-pg","name":"demo-postgres","uid":"8d1eb43f-56cb-4756-b8b4-445b3cb3bdc8","apiVersion":"kubegres.reactive-tech.io/v1","resourceVersion":"67608926"}, "reason": "StatefulSetOperation", "message": "The Spec is NOT up-to-date for a StatefulSet. 'StatefulSet name': demo-postgres-3, 'SpecName': StorageClassSize, 'Expected': 300Mi, 'Current': 200Mi"}
```

**Comment by alex-arica** [comment]
Thank you so much for checking. The details that you have provided are very useful.

It seems like after the PVC is set to a new size, the statefulset 3 and pod 3 are restated but they didn't pickup the new storage size from PVC.

The documentation that I checked recommended that we delete the statefulset without deleting the pod and recreate the statefulset that will force to pickup the storage size from PVC. 

Unfortunately, Kubernetes seems to not have an easier way when expending storage.

I will try that approach and keep you updated.

**Comment by ylck** [comment]
> Thank you so much for checking. The details that you have provided are very useful.
> 
> It seems like after the PVC is set to a new size, the statefulset 3 and pod 3 are restated but they didn't pickup the new storage size from PVC.
> 
> The documentation that I checked recommended that we delete the statefulset without deleting the pod and recreate the statefulset that will force to pickup the storage size from PVC.
> 
> Unfortunately, Kubernetes seems to not have an easier way when expending storage.
> 
> I will try that approach and keep you updated.

I see that the mongo operator's usage scenario for expansion is also to delete the statefulset first 

 https://docs.mongodb.com/kubernetes-operator/master/tutorial/resize-pv-storage/

**Comment by alex-arica** [comment]
After further research, what the operator needs to do for the StatefulSet to pickup the new storage size from the PVC, are:
- delete each StatefulSet with the option --cascade=orphan
- recreate each StatefulSet
- restart the pods associated with each StatefulSet

Unfortunately, Kubegres operator will need a major change to allow deleting the StatefulSets and recreating them. It is possible and I can give it a try but in same time I feel like I am trying to address an issue within Kubernetes codes rather than Kubegres codes.

Currently, Kubernetes is unable to automatically update a StatefulSet when a PVC is updated. 

I think it is better and cleaner to wait on the release of the Kubernetes enhancement https://github.com/kubernetes/enhancements/pull/2842 which aims in addressing this issue by allowing support for volume expansion through StatefulSets.

There are also good links about the manual steps to follow:
- https://serverfault.com/questions/955293/how-to-increase-disk-size-in-a-stateful-set
- https://www.giffgaff.io/tech/resizing-statefulset-persistent-volumes-with-zero-downtime

Please let me know about your thoughts.

**Comment by alex-arica** [comment]
From Kubegres version 1.11, we disabled the possibility of updating the size of an existing Postgres database from Kubegres (expanding storage). To make that feature work with Kubegres the changes would be too much.

I created a new ticket #49 to implement such feature once Kubernetes allow the support for volume expansion through StatefulSets.
